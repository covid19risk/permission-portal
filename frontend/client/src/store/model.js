import { types, cast, onSnapshot } from 'mobx-state-tree'
import 'mobx-react-lite/batchingForReactDom'
import Logging from '../util/logging'

const User = types
  .model({
    isSignedIn: types.maybe(types.boolean),
    email: types.string,
    isAdmin: types.boolean,
    isSuperAdmin: types.boolean,
    disabled: types.boolean,
    prefix: types.maybe(types.string),
    imageBlob: types.maybeNull(types.string),
    firstName: types.string,
    lastName: types.string,
    organizationID: types.string,
    isFirstTimeUser: types.maybe(types.boolean),
  })
  .actions((self) => {
    const __update = (updates) => {
      Object.keys(updates).forEach((key) => {
        if (self.hasOwnProperty(key)) self[key] = updates[key] // eslint-disable-line no-prototype-builtins
      })
      Logging.log('Updated User:')
      Logging.log(self)
    }

    return { __update }
  })

const Organization = types
  .model({
    id: types.string,
    name: types.string,
    welcomeText: types.string,
    enableExposureText: types.string,
    recommendExposureText: types.string,
    notifyingOthersText: types.string,
    exposureInfoText: types.string,
    exposureAboutText: types.string,
    exposureDetailsText: types.string,
    exposureDetailsLearnText: types.string,
    verificationStartText: types.string,
    verificationIdentifierText: types.string,
    verificationIdentifierAboutText: types.string,
    verificationAdministrationDateText: types.string,
    verificationReviewText: types.string,
    verificationSharedText: types.string,
    verificationNotSharedText: types.string,
    diagnosisText: types.string,
    exposureText: types.string,
    currentPageOfMembers: types.array(User),
  })
  .actions((self) => {
    const __update = (updates) => {
      Object.keys(updates).forEach((key) => {
        if (self.hasOwnProperty(key)) self[key] = updates[key] // eslint-disable-line no-prototype-builtins
      })
      Logging.log('Updated Organization:')
      Logging.log(self)
    }

    const __setCurrentPageOfMembers = (pageOfMembers) => {
      self.currentPageOfMembers = cast(pageOfMembers)
    }

    return { __update, __setCurrentPageOfMembers }
  })

const Store = types.model({
  user: User,
  organization: Organization,
})

const defaultUser = {
  isSignedIn: false,
  email: '',
  isAdmin: false,
  isSuperAdmin: false,
  disabled: false,
  prefix: '',
  firstName: '',
  lastName: '',
  organizationID: '',
  imageBlob: null,
}

const defaultOrganization = {
  id: '',
  name: '',
  welcomeText: '',
  enableExposureText: '',
  recommendExposureText: '',
  notifyingOthersText: '',
  exposureInfoText: '',
  exposureAboutText: '',
  exposureDetailsText: '',
  exposureDetailsLearnText: '',
  verificationStartText: '',
  verificationIdentifierText: '',
  verificationIdentifierAboutText: '',
  verificationAdministrationDateText: '',
  verificationReviewText: '',
  verificationSharedText: '',
  verificationNotSharedText: '',
  diagnosisText: '',
  exposureText: '',
  currentPageOfMembers: [],
}

const defaultStore = {
  user: defaultUser,
  organization: defaultOrganization,
}

let initialStore = defaultStore

// Based on https://egghead.io/lessons/react-store-store-in-local-storage
if (localStorage.getItem('store')) {
  initialStore = JSON.parse(localStorage.getItem('store'))
}

const rootStore = Store.create({
  ...initialStore,
})

// Based on https://egghead.io/lessons/react-store-store-in-local-storage
onSnapshot(rootStore, (snapshot) => {
  localStorage.setItem('store', JSON.stringify(snapshot))
})

export { rootStore, defaultUser, defaultOrganization }
