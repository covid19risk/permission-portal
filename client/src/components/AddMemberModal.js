import React from 'react';

const AddMemberModel = () => {
  return (
    <div className="loginContainer">
      <label for="name">Name</label>
      <input type="name" id="name" name="name" />
      <label for="password">Role</label>
      <input type="password" id="password" name="password" />
      <button>Submit</button>
    </div>
  );
};

export default AddMemberModel;