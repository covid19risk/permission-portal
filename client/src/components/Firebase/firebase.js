import app from 'firebase/app';
import "firebase/auth";
import * as firebaseConfigLocal from '../../config/firebase.config.local';
import * as firebaseConfigDev from '../../config/firebase.config.dev';
import * as firebaseConfigTest from '../../config/firebase.config.test';
import * as firebaseConfigProd from '../../config/firebase.config.prod';
import "firebase/firestore";
import "firebase/database";


var firebaseConfigMap = { development: firebaseConfigDev , test: firebaseConfigTest, prod: firebaseConfigProd, local: firebaseConfigLocal};

var firebaseConfigFile;
console.log(`environment is : ${process.env.NODE_ENV}`);
var key = 'dev';
if (process.env) {
  key = process.env.NODE_ENV;
}
console.log (`firebase configuration in ${key}`)
var config = firebaseConfigMap[key];

class Firebase {
  constructor() {
    app.initializeApp(config);
    this.auth = app.auth();
    this.firestore = app.firestore();
    this.db = app.database();
  }
  doCreateUserWithEmailAndPassword = (email, password) =>
    this.auth.createUserWithEmailAndPassword(email, password);

  doSignInWithEmailAndPassword = (email, password) =>
    this.auth.signInWithEmailAndPassword(email, password);
  
  doSignOut = () => this.auth.signOut();

  doPasswordReset = email => this.auth.sendPasswordResetEmail(email);

  doPasswordUpdate = password =>
    this.auth.currentUser.updatePassword(password);

  generateUserDocument = async (user, additionalData) => {
    if (!user) return;
  
    const userRef = this.firestore.doc(`users/${user.email}`);
    const snapshot = await userRef.get();
  
    
    return this.getUserDocument(user.email);
  };
    
  getUserDocument = async uid => {
    if (!uid) return null;
    try {
      const userDocument = await this.firestore.doc(`users/${uid}`).get();
  
      return {
        uid,
        ...userDocument.data()
      };
    } catch (error) {
      console.error("Error fetching user", error);
    }
  };

  getOrganizationDocument = async orgid => {
    if (!orgid) return null;
    try {
      const orgDocument = await this.firestore.doc(`organizations/${orgid}`).get();

      return orgDocument.data()
    } catch (err) {
      console.error("Error fetching organization", error);
    }
  };

  onAuthUserListener = (next, fallback) =>
    this.auth.onAuthStateChanged(authUser => {
      if (authUser) {
             this.getUserDocument(authUser.email).then(userDoc => {
                if (!userDoc.roles) {
                  userDoc.roles = {
                    ADMIN: userDoc.isAdmin,
                    SUPER_ADMIN: userDoc.isSuperAdmin
                  }
                  
                }
                // merge auth and db user
                authUser = {
                  uid: authUser.uid,
                  email: authUser.email,
                  emailVerified: authUser.emailVerified,
                  providerData: authUser.providerData,
                  ...userDoc,
                };
                next(authUser);
              });
              
      } else {
        fallback();
      }
    });

  // *** User API ***

  user = uid => this.db.ref(`/users/${uid}`);

  users = () => this.db.ref('/users');

}
export default new Firebase();