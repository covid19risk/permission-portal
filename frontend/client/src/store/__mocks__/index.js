import { rootStore } from '../model'
import React from 'react'

const RootStoreContext = React.createContext()

const createStore = (WrappedComponent) => {
  rootStore.user.__update({ isSignedIn: true, isAdmin: false, isFirstTimeUser: false })
  return class extends React.Component {
    displayName = 'newStore'
    constructor(props) {
      super(props)
      this.data = rootStore
    }
    isSignInWithEmailLink = () => {}
    render() {
      return (
        <RootStoreContext.Provider value={this}>
          <WrappedComponent {...this.props} />
        </RootStoreContext.Provider>
      )
    }
  }
}

const withStore = (WrappedComponent) => {
  return class extends React.Component {
    displayName = 'storeConsumer'
    render() {
      return (
        <RootStoreContext.Consumer>
          {(store) => {
            return <WrappedComponent store={store} {...this.props} />
          }}
        </RootStoreContext.Consumer>
      )
    }
  }
}

export { createStore, withStore }
