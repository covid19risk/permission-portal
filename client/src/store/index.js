import { types, flow } from 'mobx-state-tree'
import { firebase } from '../components/Firebase'

const User = types
  .model({
    uid: types.string,
    isAdmin: types.boolean,
    isSuperAdmin: types.boolean,
    prefix: types.string,
    firstName: types.string,
    lastName: types.string,
    role: types.maybeNull(types.string),
    organizationID: types.string
  });

const Organization = types
  .model({
    name: types.string,
    diagnosisText: types.string,
    exposureText: types.string
  })

const Store = types
  .model({
    user: types.maybeNull(User),
    organization: types.maybeNull(Organization)
  })
  .actions((self) => {
    const signIn = flow(function* (email, password) {
      try {
        yield firebase.doSignInWithEmailAndPassword(email, password)
        console.log('Successfully logged in')

        yield firebase.auth.currentUser.getIdTokenResult(true)
        const userDoc = yield firebase.getUserDocument(email)

        self.user = userDoc
        console.log("user: ", self.user)

        const orgDoc = yield firebase.getOrganizationDocument(self.user.organizationID)

        self.organization = orgDoc
        console.log("organization: ", self.organization)
      } catch (e) {
        console.log(e)
      }
    })

    const signOut = flow(function* () {
      try {
        yield firebase.doSignOut()
        console.log('Successfully logged out')
        self.user.isAdmin = false
      } catch (e) {
        console.log(e)
      }
    })

    const setOrganizationalBranding = flow(function* (diagnosisMessage, exposureMessage) {
      console.log('Setting organization branding');
      console.log(diagnosisMessage);
      console.log(exposureMessage);
      self.organization.diagnosisMessage = diagnosisMessage;
      self.organization.exposureText = exposureMessage;
      console.log('Succeeded');
    });

    return {
      signIn,
      signOut,
      setOrganizationalBranding
    }
  })

export default Store.create({
  user: null,
  organization: null
})
