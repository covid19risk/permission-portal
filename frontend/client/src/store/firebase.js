import app from 'firebase/app'
import 'firebase/auth'
import 'firebase/firestore'
import 'firebase/database'
import 'firebase/functions'
import 'firebase/analytics'
import * as firebaseConfigLocal from '../config/firebase.config.local'
import * as firebaseConfigDev from '../config/firebase.config.dev'
import * as firebaseConfigTest from '../config/firebase.config.test'
import * as firebaseConfigProd from '../config/firebase.config.prod'
import * as firebaseConfigStaging from '../config/firebase.config.staging'

var firebaseConfigMap = {
  development: firebaseConfigDev,
  test: firebaseConfigTest,
  production: firebaseConfigProd,
  local: firebaseConfigLocal,
  staging: firebaseConfigStaging,
}

const config = firebaseConfigMap[process.env ? process.env.NODE_ENV : 'development']

// declaring this as a global variable for use in index.js for Sentry environment labeling like "development", "staging", and "production"
window.env = process.env.NODE_ENV

app.initializeApp(config)
const auth = app.auth()
const SESSION = app.auth.Auth.Persistence.SESSION
const NONE = app.auth.Auth.Persistence.NONE
const db = app.firestore()
const analytics = app.analytics()
const createUserCallable = app.functions().httpsCallable('createUser')
const initiatePasswordRecoveryCallable = app.functions().httpsCallable('initiatePasswordRecovery')
const getVerificationCodeCallable = app.functions().httpsCallable('getVerificationCode')

export {
  auth,
  db,
  SESSION,
  NONE,
  analytics,
  createUserCallable,
  initiatePasswordRecoveryCallable,
  getVerificationCodeCallable,
}
