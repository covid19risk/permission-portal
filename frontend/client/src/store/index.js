import React from 'react'
import { rootStore, defaultUser, defaultOrganization } from './model'
import Logging from '../util/logging'
import { auth, db, SESSION, createUserCallable, initiatePasswordRecoveryCallable } from './firebase'

const PAGE_SIZE = 15

const RootStoreContext = React.createContext()

const createStore = (WrappedComponent) => {
  return class extends React.Component {
    constructor(props) {
      super(props)
      auth.setPersistence(SESSION)
      this.data = rootStore
      this.__userDocumentListener = null
      this.__userImageListener = null
      this.__organizationDocumentListener = null
      this.__pageOfMembersListener = null
      this.__lastVisibleMember = null // pagination helper
      this.__firstVisibleMember = null // pagination helper
      this.__signedInWithEmailLink = false // firebase doesn't tell us this so we need to track it ourself
      this.__authStateListener = auth.onAuthStateChanged(async (user) => {
        if (user) {
          Logging.log('User signed in')
          // get user's document from the db and at the same time set up a listener to respond to document changes
          if (this.__userDocumentListener === null) {
            this.__userDocumentListener = db
              .collection('users')
              .doc(user.email)
              .onSnapshot((updatedUserDocumentSnapshot) => {
                // update the store with data from this document and set isSignedIn to true
                this.data.user.__update({
                  ...updatedUserDocumentSnapshot.data(),
                  email: updatedUserDocumentSnapshot.id,
                  isSignedIn: true,
                  signedInWithEmailLink: this.__signedInWithEmailLink,
                })
                // If DNE, set up organization listener within this callback,
                // since it relies on this.data.user.organizationID being set
                if (this.__organizationDocumentListener === null) {
                  this.__organizationDocumentListener = db
                    .collection('organizations')
                    .doc(this.data.user.organizationID)
                    .onSnapshot((updatedOrganizationDocumentSnapshot) => {
                      Logging.log('Remote organization document changed')
                      this.data.organization.__update({
                        ...updatedOrganizationDocumentSnapshot.data(),
                        id: updatedOrganizationDocumentSnapshot.id,
                        currentPage: this.data.organization.currentPage,
                      })
                      if (this.data.user.isAdmin && this.__pageOfMembersListener === null) {
                        var newListener = db
                          .collection('users')
                          .where('organizationID', '==', this.data.organization.id)
                          .orderBy('lastName')
                          .orderBy('firstName')
                          .limit(PAGE_SIZE)
                          .onSnapshot((pageOfMembersSnapshot) => {
                            this.__updatePageOfMembersOnSnapshot(pageOfMembersSnapshot.docs, newListener)
                          })
                      }
                    })
                }
              })
          }
          // Get and set up listener for user's image
          if (this.__userImageListener === null) {
            this.__userImageListener = db
              .collection('userImages')
              .doc(user.email)
              .onSnapshot((updatedUserImageDocumentSnapshot) => {
                if (updatedUserImageDocumentSnapshot.exists) {
                  this.data.user.__update({
                    imageBlob: updatedUserImageDocumentSnapshot.data().blob,
                  })
                }
              })
          }
        } else {
          Logging.log('User signed out')
          // signed out
          // reset to default state
          this.data.user.__update(defaultUser)
          this.data.organization.__update(defaultOrganization)

          // detach listeners
          if (this.__userDocumentListener !== null) {
            this.__userDocumentListener()
            this.__userDocumentListener = null
          }
          if (this.__userImageListener !== null) {
            this.__userImageListener()
            this.__userImageListener = null
          }
          if (this.__organizationDocumentListener !== null) {
            this.__organizationDocumentListener()
            this.__organizationDocumentListener = null
          }
          if (this.__pageOfMembersListener !== null) {
            this.__pageOfMembersListener()
            this.__pageOfMembersListener = null
          }
        }
      })
    }

    __updatePageOfMembersOnSnapshot(pageOfMembersSnapshotDocs, newListener) {
      if (pageOfMembersSnapshotDocs.length > 0) {
        if (this.__pageOfMembersListener != newListener) {
          this.__pageOfMembersListener = newListener
        }
        // Based on https://firebase.google.com/docs/firestore/query-data/query-cursors#paginate_a_query
        this.__firstVisibleMember = pageOfMembersSnapshotDocs[0]
        this.__lastVisibleMember = pageOfMembersSnapshotDocs[pageOfMembersSnapshotDocs.length - 1]
        this.data.organization.__setCurrentPageOfMembers(
          // See https://stackoverflow.com/a/24806827
          pageOfMembersSnapshotDocs.reduce((result, userDoc) => {
            if (userDoc.id !== this.data.user.email) {
              result.push({ ...userDoc.data(), email: userDoc.id })
            }
            return result
          }, [])
        )
      }
    }

    nextPageOfMembers() {
      var newListener = db
        .collection('users')
        .where('organizationID', '==', this.data.organization.id)
        .orderBy('lastName')
        .orderBy('firstName')
        .startAfter(this.__lastVisibleMember)
        .limit(PAGE_SIZE)
        .onSnapshot((pageOfMembersSnapshot) => {
          this.__updatePageOfMembersOnSnapshot(pageOfMembersSnapshot.docs, newListener)
        })
    }

    previousPageOfMembers() {
      var newListener = db
        .collection('users')
        .where('organizationID', '==', this.data.organization.id)
        .orderBy('lastName', 'desc')
        .orderBy('firstName', 'desc')
        .startAfter(this.__firstVisibleMember)
        .limit(PAGE_SIZE)
        .onSnapshot((pageOfMembersSnapshot) => {
          this.__updatePageOfMembersOnSnapshot(pageOfMembersSnapshot.docs.reverse(), newListener)
        })
    }

    async updateUser(updates) {
      try {
        await db.collection('users').doc(this.data.user.email).update(updates)
      } catch (err) {
        Logging.error('Error updating users', err)
      }
    }

    async updateUserImage(blob) {
      try {
        // .set() with { merge: true } so that if the document dne, it's created, otherwise its updated
        await db.collection('userImages').doc(this.data.user.email).set({ blob: blob }, { merge: true })
      } catch (err) {
        Logging.error('Error updating image', err)
      }
    }

    async updateOrganization(updates) {
      try {
        await db.collection('organizations').doc(this.data.organization.id).update(updates)
      } catch (err) {
        Logging.error('Error updating organization texts', err)
        throw err
      }
    }

    async signInWithEmailAndPassword(email, password) {
      this.__signedInWithEmailLink = false
      await auth.signInWithEmailAndPassword(email, password)
    }

    async signInWithEmailLink(email, link) {
      this.__signedInWithEmailLink = true
      return auth.signInWithEmailLink(email, link)
    }

    async signOut() {
      await auth.signOut()
    }

    async updatePassword(password) {
      await auth.currentUser.updatePassword(password)
    }

    async createUser(newUser) {
      try {
        const result = await createUserCallable(newUser)
        Logging.log(`Created new user: ${JSON.stringify(result.data)}`)
        return result.data
      } catch (err) {
        Logging.log(err)
        throw err
      }
    }

    async sendPasswordResetEmail(email) {
      try {
        await auth.sendPasswordResetEmail(email)
        return true
      } catch (err) {
        Logging.error(err)
        throw err
      }
    }

    async sendPasswordRecoveryEmail(email) {
      try {
        await initiatePasswordRecoveryCallable({ email: email })
        // Save the email locally so you don't need to ask the user for it again if they open the link on the same device.
        // See https://firebase.google.com/docs/auth/web/email-link-auth#send_an_authentication_link_to_the_users_email_address
        window.localStorage.setItem('emailForSignIn', email)
        return true
      } catch (err) {
        Logging.error(err)
        throw err
      }
    }

    async updateUserByEmail(email, updates) {
      try {
        await db.collection('users').doc(email).update(updates)
      } catch (err) {
        Logging.error(`Error updating user: ${email}`, err)
        throw err
      }
    }

    isSignInWithEmailLink(link) {
      return auth.isSignInWithEmailLink(link)
    }

    setPasswordResetCompletedInCurrentSession(val) {
      this.data.user.__update({ passwordResetCompletedInCurrentSession: val })
    }

    displayName = 'storeProvider'

    render() {
      return (
        <RootStoreContext.Provider value={this}>
          <WrappedComponent {...this.props} />
        </RootStoreContext.Provider>
      )
    }
  }
}

const withStore = (WrappedComponent) => {
  return class extends React.Component {
    displayName = 'storeConsumer'
    render() {
      return (
        <RootStoreContext.Consumer>
          {(store) => <WrappedComponent store={store} {...this.props} />}
        </RootStoreContext.Consumer>
      )
    }
  }
}

export { createStore, withStore, PAGE_SIZE }
