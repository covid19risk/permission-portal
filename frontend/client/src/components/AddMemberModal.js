import React, { useState } from 'react'
import PendingOperationButton from '../components/PendingOperationButton'
import RoleSelector from '../components/RoleSelector'
import Modal from '../components/Modal'
import ModalInput from '../components/ModalInput'
import * as ROLES from '../constants/roles'
import { withStore } from '../store'
import { observer } from 'mobx-react'
import validateEmail from '../util/validateEmail'
import '../../styles/screens/add_member_modal.scss' // NOTE: see note in index.scss

const ValidationResult = (succeeded, failureReason) => {
  return {
    succeeded: succeeded,
    failureReason: failureReason,
  }
}

const ValidationRules = [
  (value) => {
    // firstName
    return ValidationResult(value.length > 0, 'First name required')
  },
  (value) => {
    // lastName
    return ValidationResult(value.length > 0, 'Last name required')
  },
  (value) => {
    // email
    return ValidationResult(validateEmail(value), 'Email address must be valid')
  },
  () => {
    // role
    return ValidationResult(true, 'Not possible')
  },
]

const AddMemberModalBase = observer((props) => {
  const [state, setState] = useState({
    firstName: '',
    firstNameValidationFailed: false,
    firstNameValidationMessage: '',
    lastName: '',
    lastNameValidationFailed: false,
    lastNameValidationMessage: '',
    email: '',
    emailValidationFailed: false,
    emailValidationMessage: '',
    role: ROLES.NON_ADMIN_LABEL,
    roleValidationFailed: false,
    roleValidationMessage: '',
  })

  const tryCreateUser = () => {
    let newState = {}
    let validationSucceeded = true
    const values = [state.firstName, state.lastName, state.email, state.role]
    ValidationRules.forEach((validationRule, index) => {
      const { succeeded, failureReason } = validationRule(values[index])
      if (succeeded === false) {
        if (index === 0) {
          newState.firstNameValidationFailed = true
          newState.firstNameValidationMessage = failureReason
        } else if (index === 1) {
          newState.lastNameValidationFailed = true
          newState.lastNameValidationMessage = failureReason
        } else if (index === 2) {
          newState.emailValidationFailed = true
          newState.emailValidationMessage = failureReason
        } else if (index === 3) {
          newState.roleValidationFailed = true
          newState.roleValidationMessage = failureReason
        }
        validationSucceeded = false
      }
    })

    if (validationSucceeded === false) {
      setState({ ...state, ...newState })
      return
    }

    return props.store
      .createUser({
        email: state.email,
        firstName: state.firstName,
        lastName: state.lastName,
        isAdmin: state.role === ROLES.ADMIN_LABEL,
      })
      .then(props.onSuccess, props.onFailure)
  }

  function handleChange(e) {
    if (e.target.name === 'firstName') {
      setState({ ...state, firstName: e.target.value })
    } else if (e.target.name === 'lastName') {
      setState({ ...state, lastName: e.target.value })
    } else if (e.target.name === 'email') {
      setState({ ...state, email: e.target.value })
    } else {
      setState({ ...state, role: e.target.value })
    }
  }

  return (
    <Modal title="Add Member" hidden={props.hidden} onClose={props.onClose} containerClass="add-member-modal-container">
      <form className="modal-form add-member-form">
        <ModalInput
          label="First Name"
          id="firstName"
          required={true}
          value={state.firstName}
          onChange={handleChange}
          validation={state.firstNameValidationFailed}
          validationMessage={state.firstNameValidationMessage}
        />

        <ModalInput
          label="Last Name"
          id="lastName"
          required={true}
          value={state.lastName}
          onChange={handleChange}
          validation={state.lastNameValidationFailed}
          validationMessage={state.lastNameValidationMessage}
        />

        <ModalInput
          label="Email"
          id="email"
          required={true}
          value={state.email}
          onChange={handleChange}
          validation={state.emailValidationFailed}
          validationMessage={state.emailValidationMessage}
        />

        <div className="modal-input">
          <label htmlFor="role" style={{ marginTop: state.emailValidationFailed ? '6px' : null }}>
            Role<span>*</span>
          </label>
          <RoleSelector isAdmin={false} id="role" required={true} onChange={handleChange} />
          {state.roleValidationFailed && <div className="validationResult">{state.roleValidationMessage}</div>}
        </div>
        <PendingOperationButton className="save-button" operation={tryCreateUser}>
          Submit
        </PendingOperationButton>
      </form>
    </Modal>
  )
})

const AddMemberModal = withStore(AddMemberModalBase)

export default AddMemberModal
