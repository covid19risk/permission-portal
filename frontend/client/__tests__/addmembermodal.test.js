import { mount, shallow } from 'enzyme'
import React from 'react'
import AddMemberModal from '../src/components/AddMemberModal'
import { createStore } from '../src/store'
import { createUserCallable } from '../src/store/firebase'
import toJson from 'enzyme-to-json'

jest.mock('../src/store/firebase', () => {
  return {
    ...jest.requireActual('../src/store/firebase'),
    createUserCallable: jest.fn(async (newUser) => newUser),
  }
})

describe('AddMemberModal', () => {
  let AddMemberModalWrapped
  let wrapper
  let onAddMemberCancel

  beforeEach(() => {
    AddMemberModalWrapped = createStore(AddMemberModal)

    var showAddMemberModal = true
    onAddMemberCancel = jest.fn()
    const onAddMemberSuccess = jest.fn()
    const onAddMemberFailure = jest.fn()

    wrapper = mount(
      <AddMemberModalWrapped
        hidden={!showAddMemberModal}
        onClose={onAddMemberCancel}
        onSuccess={onAddMemberSuccess}
        onFailure={onAddMemberFailure}
      />
    )
  })

  afterEach(() => {
    wrapper.unmount()
  })

  it('renders correctly', () => {
    var showAddMemberModal = true
    const onAddMemberCancel = jest.fn()
    const onAddMemberSuccess = jest.fn()
    const onAddMemberFailure = jest.fn()

    const wrapper = shallow(
      <AddMemberModalWrapped
        hidden={!showAddMemberModal}
        onClose={onAddMemberCancel}
        onSuccess={onAddMemberSuccess}
        onFailure={onAddMemberFailure}
      />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('closes when close button is clicked', () => {
    wrapper.find('.close-btn').at(0).simulate('click')
    expect(onAddMemberCancel.mock.calls.length).toBe(1)
  })

  it('closes when background is clicked', () => {
    wrapper.find('.modal-background').at(0).simulate('click')
    expect(onAddMemberCancel.mock.calls.length).toBe(1)
  })

  it.each([
    ['', 'test', '', ''],
    ['test', 'test', 'invalidemail', 'Test Validator'],
  ])('does not allow invalid input', async (firstName, lastName, email, role) => {
    wrapper
      .find('#firstName')
      .at(1)
      .simulate('change', {
        target: {
          name: 'firstName',
          value: firstName,
        },
      })
    wrapper
      .find('#lastName')
      .at(1)
      .simulate('change', {
        target: {
          name: 'lastName',
          value: lastName,
        },
      })
    wrapper
      .find('#email')
      .at(1)
      .simulate('change', {
        target: {
          name: 'email',
          value: email,
        },
      })
    wrapper
      .find('#role')
      .at(1)
      .simulate('change', {
        target: {
          name: 'role',
          value: role,
        },
      })

    wrapper.find('.button').at(0).simulate('click')

    await global.waitForComponentToPaint(wrapper)

    expect(createUserCallable).toHaveBeenCalledTimes(0)
  })

  it('creates new user succesfully', async () => {
    const firstName = 'Test'
    const lastName = 'Test'
    const email = 'testuser@soylentgreen.com'
    const role = 'Test Validator'

    wrapper
      .find('#firstName')
      .at(1)
      .simulate('change', {
        target: {
          name: 'firstName',
          value: firstName,
        },
      })
    wrapper
      .find('#lastName')
      .at(1)
      .simulate('change', {
        target: {
          name: 'lastName',
          value: lastName,
        },
      })
    wrapper
      .find('#email')
      .at(1)
      .simulate('change', {
        target: {
          name: 'email',
          value: email,
        },
      })
    wrapper
      .find('#role')
      .at(1)
      .simulate('change', {
        target: {
          name: 'role',
          value: role,
        },
      })

    wrapper.find('.button').at(0).simulate('click')

    await global.waitForComponentToPaint(wrapper)

    expect(createUserCallable).toBeCalledWith({
      email: email,
      firstName: firstName,
      lastName: lastName,
      isAdmin: false,
    })
  })
})
