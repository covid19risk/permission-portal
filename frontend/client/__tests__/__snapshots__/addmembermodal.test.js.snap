// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMemberModal renders correctly 1`] = `
<ContextProvider
  value={
    _temp {
      "__authStateListener": [Function],
      "__firstVisibleMember": null,
      "__lastVisibleMember": null,
      "__membersListener": null,
      "__organizationDocumentListener": null,
      "__signedInWithEmailLink": false,
      "__userDocumentListener": null,
      "__userImageListener": null,
      "analytics": Object {},
      "context": Object {},
      "data": Object {
        "organization": Object {
          "diagnosisText": "",
          "enableExposureText": "",
          "exposureAboutText": "",
          "exposureDetailsLearnText": "",
          "exposureDetailsText": "",
          "exposureInfoText": "",
          "exposureText": "",
          "id": "",
          "logoBlob": null,
          "members": Array [],
          "membersPage": 1,
          "name": "",
          "notifyingOthersText": "",
          "recommendExposureText": "",
          "verificationAdministrationDateText": "",
          "verificationIdentifierAboutText": "",
          "verificationIdentifierText": "",
          "verificationNotSharedText": "",
          "verificationReviewText": "",
          "verificationSharedText": "",
          "verificationStartText": "",
          "welcomeText": "",
        },
        "user": Object {
          "disabled": false,
          "email": "",
          "firstName": "",
          "imageBlob": null,
          "isAdmin": false,
          "isFirstTimeUser": true,
          "isSignedIn": false,
          "lastName": "",
          "organizationID": "",
          "passwordResetCompletedInCurrentSession": false,
          "passwordResetRequested": false,
          "prefix": "",
          "signedInWithEmailLink": false,
        },
      },
      "displayName": "storeProvider",
      "props": Object {
        "hidden": false,
        "onClose": [MockFunction],
        "onFailure": [MockFunction],
        "onSuccess": [MockFunction],
      },
      "refs": Object {},
      "setState": [Function],
      "state": null,
      "updater": Updater {
        "_callbacks": Array [],
        "_renderer": ReactShallowRenderer {
          "_context": Object {},
          "_didScheduleRenderPhaseUpdate": false,
          "_dispatcher": Object {
            "readContext": [Function],
            "useCallback": [Function],
            "useContext": [Function],
            "useDebugValue": [Function],
            "useDeferredValue": [Function],
            "useEffect": [Function],
            "useImperativeHandle": [Function],
            "useLayoutEffect": [Function],
            "useMemo": [Function],
            "useReducer": [Function],
            "useRef": [Function],
            "useResponder": [Function],
            "useState": [Function],
            "useTransition": [Function],
          },
          "_element": <_temp
            hidden={false}
            onClose={[MockFunction]}
            onFailure={[MockFunction]}
            onSuccess={[MockFunction]}
          />,
          "_firstWorkInProgressHook": null,
          "_forcedUpdate": false,
          "_instance": [Circular],
          "_isReRender": false,
          "_newState": null,
          "_numberOfReRenders": 0,
          "_renderPhaseUpdates": null,
          "_rendered": <Context.Provider
            value={[Circular]}
          >
            <_temp2
              hidden={false}
              onClose={[MockFunction]}
              onFailure={[MockFunction]}
              onSuccess={[MockFunction]}
            />
          </Context.Provider>,
          "_rendering": false,
          "_updater": [Circular],
          "_workInProgressHook": null,
        },
      },
    }
  }
>
  <_temp2
    hidden={false}
    onClose={[MockFunction]}
    onFailure={[MockFunction]}
    onSuccess={[MockFunction]}
  />
</ContextProvider>
`;
